src/network.rs

```diff
@@ -0,0 +1,358 @@
+    // hand the response off to the error-handler, continue on success
+    error_handler(&mut res, eh_rb, &hvalue)?;
+    // Write the initial read length value into the pointer
+    let contentlength = res.content_length().ok_or(ProjError::ContentLength)? as usize;
+    unsafe { out_size_read.write(contentlength) };
+    let headers = res.headers().clone();
+    // Copy the downloaded bytes into the buffer so it can be passed around
+    unsafe {
+        &res.bytes()?
+            .as_ptr()
+            .copy_to_nonoverlapping(buffer as *mut u8, contentlength.min(size_to_read))
+    };
+    // Store req into the handle so new ranges can be queried
+    let hd = HandleData::new(req, headers, None);
+    // heap-allocate the struct and cast it to a void pointer so it can be passed around to PROJ
+    let hd_boxed = Box::new(hd);
+    let void: *mut c_void = Box::into_raw(hd_boxed) as *mut c_void;
+    let opaque: *mut PROJ_NETWORK_HANDLE = void as *mut PROJ_NETWORK_HANDLE;
+    // If everything's OK, set the error string to empty
+    let err_string = "";
+    unsafe {
+        out_error_string.copy_from_nonoverlapping(err_string.as_ptr().cast(), err_string.len());
+        out_error_string.add(err_string.len()).write(0);
+    }
+    Ok(opaque)
+}
+
+/// Network callback: close connection and drop handle data (client and headers)
+pub(crate) unsafe extern "C" fn network_close(
+    _: *mut PJ_CONTEXT,
+    handle: *mut PROJ_NETWORK_HANDLE,
+    _: *mut c_void,
+) {
+    // Reconstitute the Handle data so it can be dropped
+    let hd = &*(handle as *const c_void as *mut HandleData);
+    // Reconstitute and drop the header value returned by network_get_header_value,
+    // since PROJ never explicitly returns it to us
+    if let Some(header) = hd.hptr {
+        let _ = CString::from_raw(header as *mut i8);
+    }
+    let _ = *hd;
+}
+
+/// Network callback: get header value
+///
+/// Note that this function is a facade for _network_get_header_value
+pub(crate) unsafe extern "C" fn network_get_header_value(
+    pc: *mut PJ_CONTEXT,
+    handle: *mut PROJ_NETWORK_HANDLE,
+    header_name: *const c_char,
+    ud: *mut c_void,
+) -> *const c_char {
+    let mut hd = &mut *(handle as *const c_void as *mut HandleData);
+    match _network_get_header_value(pc, handle, header_name, ud) {
+        Ok(res) => res,
+        Err(_) => {
+            // an empty value will cause an error upstream in libproj, which is the intention
+            let hvalue = "";
+            // unwrapping an empty str is fine
+            let cstr = CString::new(hvalue).unwrap();
+            let err = cstr.into_raw();
+            hd.hptr = Some(err);
+            err
+        }
+    }
+}
+
+/// Network callback: get header value
+fn _network_get_header_value(
+    _: *mut PJ_CONTEXT,
+    handle: *mut PROJ_NETWORK_HANDLE,
+    header_name: *const c_char,
+    _: *mut c_void,
+) -> Result<*const c_char, ProjError> {
+    let lookup = _string(header_name)?.to_lowercase();
```

Comment:
curious - why do we `to_lowercase` here? 

Is it a Reqwest thing that all headers are lowercased?

Reply:
Yep – this actually took me a solid day to figure out, because the download operation would error out after `network_open` had returned, for no discernible reason, even though it had retrieved data and written headers. Turns out that Reqwest (or possibly Hyper?) lowercases the headers. Headers are supposed to be case-insensitive, so this is a gotcha in the Hyper `HeaderMap` implementation: `get()` is case-sensitive, and maybe it shouldn't be.